File: visitor_pattern.py
from abc import ABC, abstractmethod
import ast

#Step 1: Define the expression class and it's child classes
class Expression(ABC):
    def accept(self, visitor):
        pass

    loaders = {}

    @staticmethod
    def register_loader(key, loader):
        Expression.loaders[key] = loader

    @staticmethod
    def load(s):
        key = s.read(1)
        loader = Expression.loaders.get(key)
        if loader is None:
            raise ValueError(f"Found unexpected token {key}")
        return loader(s)

class Constant(Expression):
    def __init__(self, x):
        self.x = x
        Expression.register_loader("Constant", Constant.load)

    def accept(self, visitor):
        visitor.visit_constant(self)

    @staticmethod
    def load(s):
        x = float(s.read().split()[0])
        return Constant(x)

class Variable(Expression):
    def __init__(self, name):
        self.name = name
        Expression.register_loader("Variable", Variable.load)

    def accept(self, visitor):
        visitor.visit_variable(self)

    @staticmethod
    def load(s):
        name = s.read().split()[0]
        return Variable(name)

class Parentheses(Expression):
    def __init__(self, expr):
        self.expr = expr
        Expression.register_loader("Parentheses", Parentheses.load)

    def accept(self, visitor):
        visitor.visit_parentheses(self)

    @staticmethod
    def load(s):
        expr = Expression.load(s)
        return Parentheses(expr)

class Op(Expression):
    def __init__(self, op, l, r):
        self.op = op
        self.l = l
        self.r = r
        Expression.register_loader("Op", Op.load)

    def accept(self, visitor):
        visitor.visit_op(self)

    @staticmethod
    def load(s):
        op = s.read(1)
        l = Expression.load(s)
        r = Expression.load(s)
        return Op(op, l, r)

class Visitor(ABC):
    @abstractmethod
    def visit_constant(self, constant:Constant) -> None:
        pass

    @abstractmethod
    def visit_op(self, op:Op) -> None:
        pass
        
    @abstractmethod
    def visit_variable(self, variable:Variable) -> None:
        pass
    
    @abstractmethod
    def visit_parentheses(self, parentheses:Parentheses) -> None:
        pass
class ComputeVisitor(Visitor):

    def __init__(self) -> None:
        self.variables = variables or {}
        self.result_ = 0

    def visit_constant(self, constant):
        self.result_ = constant.x

    def visit_op(self, op):
        op.l.accept(self)
        l = self.result_

        op.r.accept(self)
        r = self.result_

        operators = {
            '^': lambda x, y: x ** y,
            '+': lambda x, y: x + y,
            '-': lambda x, y: x - y,
            '*': lambda x, y: x * y,
            '/': lambda x, y: x // y,
        }

        if op.op in operators:
            self.result_ = operators[op.op](l, r)

        else:
            print("Invalid Operand", op.op)

    def visit_variable(self, variable):
        if variable.name in self.variables:
            self.result_ = self.variables[variable.name]
        else:
            print(f"Variable {variable.name} not defined.")
            self.result_ = 0

    def visit_parentheses(self, parentheses):
        parentheses.expr.accept(self)

class PrettyPrintVisitor(Visitor):
    def __init__(self) -> None:
        self.result_ = ""

    def visit_constant(self, constant):
        self.result_ = "(" + str(constant.x) + ")"

    def visit_op(self, op):
        result = []
        result.append("(")
        op.l.accept(self)
        result.append(self.result_)
        result.append(op.op)

        op.r.accept(self)
        result.append(self.result_)

        result.append(")")
        self.result_ = "".join(result)


# Step 3: Implement the parser
def parse(t: str) -> Expression:
    for i in range(len(t) - 1, -1, -1):
        if t[i] in ['+', '-']:
            return Op(t[i], parse(t[:i]), parse(t[i + 1:]))
            
    for i in range(len(t) - 1, -1, -1):
        if t[i] in ['*', '/']:
            return Op(t[i], parse(t[:i]), parse(t[i + 1:]))
            
    for i in range(len(t) - 1, -1, -1):
        if t[i] == '^':
            return Op(t[i], parse(t[:i]), parse(t[i + 1:]))
    
    return Constant(int(t))


# Step 0: Check if input string is a valid expression
def is_valid_expression(input_string):
    try:
        # Use ast.parse to parse the input string
        if not input_string or input_string == " ":
            return False
        ast.parse(input_string)
        return True
    except SyntaxError:
        return False
    
def main(s):
    # test expression
    # s = "16/4+4*3-3+1"
    is_valid = is_valid_expression(s)

    # if not true end execution
    if not is_valid:
        return "Invalid Input"

    result = {}
    # else continue execution 
    e = parse(s)
    cmp_v = ComputeVisitor()
    e.accept(cmp_v)
    print("Result =",cmp_v.result_)
    result['result'] = cmp_v.result_

    pp_v = PrettyPrintVisitor()
    e.accept(pp_v)
    print("PrettyPrint ->", pp_v.result_)
    result['pretty_print'] = pp_v.result_

    return result


if __name__ == "__main__":
    main()

File: C++/visitor_pattern.cc
#include <iostream>
#include <string>
#include <sstream>
#include <unordered_map>
#include <cmath>
#include <functional>
#include <iomanip>
#include <typeinfo>

class Visitor;

// Step 1: Define the expression class and its child classes
class Expression {
public:
    virtual ~Expression() = default;
    virtual void accept(Visitor& visitor) = 0;
};


// Step 2: Define the visitor classes
class Visitor {
public:
    virtual ~Visitor() = default;
    virtual void visit_constant(class Constant* constant) = 0;
    virtual void visit_op(class Op* op) = 0;
    virtual void visit_variable(class Variable* variable) = 0;
    virtual void visit_parentheses(class Parentheses* parentheses) = 0;
};

class Constant : public Expression {
public:
    explicit Constant(double x) : x(x) {}
    void accept(Visitor& visitor) override {
        visitor.visit_constant(this);
    }

    double x;
};

class Variable : public Expression {
public:
    explicit Variable(const std::string& name) : name(name) {}
    void accept(Visitor& visitor) override {
        visitor.visit_variable(this);
    }

    std::string name;
};

class Parentheses : public Expression {
public:
    explicit Parentheses(Expression* expr) : expr(expr) {}
    ~Parentheses() override {
        delete expr;
    }
    void accept(Visitor& visitor) override {
        visitor.visit_parentheses(this);
    }

    Expression* expr;
};

class Op : public Expression {
public:
    Op(char op, Expression* l, Expression* r) : op(op), l(l), r(r) {}
    ~Op() override {
        delete l;
        delete r;
    }
    void accept(Visitor& visitor) override {
        visitor.visit_op(this);
    }

    char op;
    Expression* l;
    Expression* r;
};

// Step 3: Implement the parser
Expression* parse(std::istringstream& iss);

Expression* parse(const std::string& t) {
    std::istringstream iss(t);
    return parse(iss);
}

Expression* parse(std::istringstream& iss) {
    int c = 0;
    std::string t;
    iss >> t;

    for (int i = t.length() - 1; i >= 0; --i) {
        if (t[i] == ')') {
            ++c;
        } else if (t[i] == '(') {
            --c;
        } else if (c == 0 && (t[i] == '+' || t[i] == '-')) {
            return new Op(t[i], parse(t.substr(0, i)), parse(t.substr(i + 1)));
        }
    }

    for (int i = t.length() - 1; i >= 0; --i) {
        if (t[i] == ')') {
            ++c;
        } else if (t[i] == '(') {
            --c;
        } else if (c == 0 && (t[i] == '*' || t[i] == '/')) {
            return new Op(t[i], parse(t.substr(0, i)), parse(t.substr(i + 1)));
        }
    }

    for (int i = t.length() - 1; i >= 0; --i) {
        if (t[i] == ')') {
            ++c;
        } else if (t[i] == '(') {
            --c;
        } else if (c == 0 && t[i] == '^') {
            return new Op(t[i], parse(t.substr(0, i)), parse(t.substr(i + 1)));
        }
    }

    if (t[0] == '(' && t.back() == ')') {
        return parse(t.substr(1, t.length() - 2));
    }

    return new Constant(std::stod(t));
}


// Step 0: Check if input string is a valid expression
bool is_valid_expression(const std::string& input_string) {
    if (input_string.empty() || input_string == " ") {
        return false;
    }
    try {
        std::istringstream iss(input_string);
        while (iss) {
            std::string token;
            iss >> token;
            if (!token.empty()) {
                // Convert to double and check for syntax errors
                std::stod(token);
            }
        }
        return true;
    } catch (const std::exception& e) {
        return false;
    }
}

// Step 4: Define ComputeVisitor and PrettyPrintVisitor
class ComputeVisitor : public Visitor {
public:
    ComputeVisitor(std::unordered_map<std::string, double>& variables) : variables(variables), result_(0) {}
    void visit_constant(Constant* constant) override {
        result_ = constant->x;
    }
    void visit_op(Op* op) override {
        op->l->accept(*this);
        double l = result_;

        op->r->accept(*this);
        double r = result_;

        static const std::unordered_map<char, std::function<double(double, double)>> operators = {
            {'^', [](double x, double y) { return std::pow(x, y); }},
            {'+', [](double x, double y) { return x + y; }},
            {'-', [](double x, double y) { return x - y; }},
            {'*', [](double x, double y) { return x * y; }},
            {'/', [](double x, double y) { return x / y; }},
        };

        auto it = operators.find(op->op);
        if (it != operators.end()) {
            result_ = it->second(l, r);
        } else {
            std::cout << "Invalid Operand " << op->op << std::endl;
        }
    }
    void visit_variable(Variable* variable) override {
        if (variables.find(variable->name) != variables.end()) {
            result_ = variables[variable->name];
        } else {
            std::cout << "Variable " << variable->name << " not defined." << std::endl;
            result_ = 0;
        }
    }
    void visit_parentheses(Parentheses* parentheses) override {
        parentheses->expr->accept(*this);
    }
    double result_;
private:
    std::unordered_map<std::string, double>& variables;
};

class PrettyPrintVisitor : public Visitor {
public:
    void visit_constant(Constant* constant) override {
        std::string numStr = std::to_string(constant->x);
        numStr = numStr.substr(0, numStr.find(".") + 3);
        result_ = "(" + numStr + ")";
    }
    void visit_op(Op* op) override {
        std::string temp;
        op->l->accept(*this);
        temp += result_;
        temp += op->op;
        op->r->accept(*this);
        result_ = "(" + temp + result_ + ")";
    }
    void visit_variable(Variable* variable) override {
        result_ = variable->name;
    }
    void visit_parentheses(Parentheses* parentheses) override {
        parentheses->expr->accept(*this);
    }
    std::string result_;
};

// Main function
int main() {
    std::string s = "3+5/2";
    std::unordered_map<std::string, double> variables;
    if (!is_valid_expression(s)) {
        std::cout << "Invalid Input" << std::endl;
        return 1;
    }
    Expression* e = parse(s);
    ComputeVisitor cmp_v(variables);
    e->accept(cmp_v);
    std::cout << "Result = " << cmp_v.result_ << std::endl;
    delete e;

    PrettyPrintVisitor pp_v;
    e = parse(s);
    e->accept(pp_v);
    std::cout << "PrettyPrint -> " << pp_v.result_ << std::endl;
    delete e;

    return 0;
}
